题目：掷三个骰子，不分先后次序，问可能出现的组合有多少种？

解法一

In [1]:
d = '123456'
_all = [i+j+k for i in d for j in d for k in d]
len(_all)

216

In [2]:
def _sort(x):
    return ''.join(sorted(list(x)))

In [3]:
_sort('324')

'234'

In [4]:
result = set(map(_sort, _all))

In [5]:
result

{'111',
 '112',
 '113',
 '114',
 '115',
 '116',
 '122',
 '123',
 '124',
 '125',
 '126',
 '133',
 '134',
 '135',
 '136',
 '144',
 '145',
 '146',
 '155',
 '156',
 '166',
 '222',
 '223',
 '224',
 '225',
 '226',
 '233',
 '234',
 '235',
 '236',
 '244',
 '245',
 '246',
 '255',
 '256',
 '266',
 '333',
 '334',
 '335',
 '336',
 '344',
 '345',
 '346',
 '355',
 '356',
 '366',
 '444',
 '445',
 '446',
 '455',
 '456',
 '466',
 '555',
 '556',
 '566',
 '666'}

In [6]:
len(result)

56

In [7]:
dic = {3:0, 2:0, 1:0}
for a in result:
    dic[len(set(list(a)))] += 1
    
print(dic)

{1: 6, 2: 30, 3: 20}


总结： 解决问题的关键是 sort。'123', '132', '213', '231', '312', '321' 这六种可能的组合在 sort 之后都是同样的 '123'。
起初，我用 set() 来判断，对于 '123' 这种情况成立，但是对于 '112' 和 '221' 这两种不同的组合，set 的结果是一样的，因此 set 判断不可用。

解法二

In [12]:
def equal(x, y):
    if set(list(x)) != set(list(y)):
        return False
    else:
        for element in set(list(x)):
            if x.count(element) != y.count(element):
                return False
        return True

In [13]:
equal('121', '112')

True

In [14]:
equal('121', '221')

False

In [15]:
d = '123456'
_all = [i+j+k for i in d for j in d for k in d]

In [30]:
temp = []
result1 = []
for elm in _all:
    if elm in temp:
        continue
    temp1 = []
    for elm1 in _all:
        if elm1 in temp:
            continue
        if equal(elm, elm1):
            temp1.append(elm1)
    temp.extend(temp1)
    result1.append(temp1)

In [31]:
len(result1)

56

In [32]:
result1

[['111'],
 ['112', '121', '211'],
 ['113', '131', '311'],
 ['114', '141', '411'],
 ['115', '151', '511'],
 ['116', '161', '611'],
 ['122', '212', '221'],
 ['123', '132', '213', '231', '312', '321'],
 ['124', '142', '214', '241', '412', '421'],
 ['125', '152', '215', '251', '512', '521'],
 ['126', '162', '216', '261', '612', '621'],
 ['133', '313', '331'],
 ['134', '143', '314', '341', '413', '431'],
 ['135', '153', '315', '351', '513', '531'],
 ['136', '163', '316', '361', '613', '631'],
 ['144', '414', '441'],
 ['145', '154', '415', '451', '514', '541'],
 ['146', '164', '416', '461', '614', '641'],
 ['155', '515', '551'],
 ['156', '165', '516', '561', '615', '651'],
 ['166', '616', '661'],
 ['222'],
 ['223', '232', '322'],
 ['224', '242', '422'],
 ['225', '252', '522'],
 ['226', '262', '622'],
 ['233', '323', '332'],
 ['234', '243', '324', '342', '423', '432'],
 ['235', '253', '325', '352', '523', '532'],
 ['236', '263', '326', '362', '623', '632'],
 ['244', '424', '442'],
 ['245', '2

解法二花了一晚上的时间才搞定。问题出在 temp1 = [ ] 的位置上，它应该置于内层 for loop 之外的，我之前把它放在内层 for loop 之内了！！！
谨记：
1. 如果计划用一个空 list 接收 for loop 中产生的数据，这个空列表必须置于 for loop 之外，否则 for loop 没执行一步， 这个列表就被清空，根本起不到接收器的作用。
2. 要提高 debug 的能力，涉及到 loop 时，把中间步骤的结果 print 出来，有助于发现问题。